scripts/get_tau_pet.py

```diff
@@ -1,94 +1,121 @@
-baseline.sort_values(by='RID', inplace=True)
-tau_df.sort_values(by=['RID', 'EXAMDATE'],inplace=True)
+    ''' Load tau data and return df '''
 
-# group subjects in the tau df, look them up in the baseline df, and if they match return the tau
-# value closest to the session date in baseline. Return a list of dfs, one per subject. This seems
-# a silly and expensive way to do this! But, I ran into all sorts of problems with the datetime format
-# and this is the only way I figured it out...
-dfs = []
-for i, grp in tau_df.groupby(level='RID'):
-    for subject, session in zip(baseline.index, baseline.session):
-        if i == subject:
-            df = grp
-            df.set_index('EXAMDATE', inplace=True)
-            tau = pd.DataFrame(df['META_TEMPORAL_SUVR'][df.index[[df.index.get_loc(session, method='nearest')]]])
-            tau['EXAMDATE'] = tau.index
-            tau.index = [subject]
-            tau.rename_axis('RID',inplace=True)
-            dfs.append(tau)
+    # load flortaucipir spreadsheet
+    tau_df = pd.read_csv(file, index_col=0, header=0, low_memory=False)
 
-# concatenate individual tau dfs
-master_tau = pd.concat(dfs)
+    # keep only relevant columns
+    tau_df = tau_df.filter(['EXAMDATE','META_TEMPORAL_SUVR'], axis=1)
 
-# merge tau data into baseline df
-baseline = baseline.join(master_tau)
+    # convert to datetime
+    tau_df['EXAMDATE'] = pd.to_datetime(tau_df['EXAMDATE'])
 
-# create a new df with only tau values that were collected within a 12month window of the baseline visit
-baseline_tau = baseline[(baseline.EXAMDATE > baseline.date_lower) & (baseline.EXAMDATE < baseline.date_upper)]
+    # sort tau df by index and date. Also throws an error
+    tau_df.sort_values(by=['RID', 'EXAMDATE'],inplace=True)
+    
+    return (tau_df)
 
-# save df
-baseline_tau.to_csv('/Users/natashaclarke/Documents/SIMEXP/cwas_neuro_sandbox/adni_cwas/baseline_tau.csv')
\ No newline at end of file
+def match_baseline_tau(baseline_df, tau_df):
+    
+    '''
+    Group subjects in the tau df, look them up in the baseline df, and if they match return the tau
+    value closest to the session date in baseline. Return a list of dfs, one per subject, merge into
+    baseline df, and then keep only those within a 12 month window. This seems a silly and expensive way 
+    to do this! But, I ran into all sorts of problems with the datetime format and this is the only way 
+    I figured it out...
+    '''
+    
+    tau_dfs_list = []
+    for tau_id, group in tau_df.groupby(level='RID'):
+        for baseline_id, session in zip(baseline_df.index, baseline_df.session):
+            if tau_id == baseline_id:
+                participant_df = group
+                participant_df.set_index('EXAMDATE', inplace=True)
+                participant_tau = pd.DataFrame(participant_df['META_TEMPORAL_SUVR'][participant_df.index[[participant_df.index.get_loc(session, method='nearest')]]])
```

Comment:
Avoiding wrapping around so many things in one line. It will be difficult to read and figure out what this line is doing in 3 months time.
In general if you see more than two same type of bracket in a cluster (i.e. `]]]` or `])]` something along these line), it would be difficult to find out which bracket match up with which starting one, it will be difficult to figure out what's it doing.

```suggestion
      nearest_something = [participant_df.index.get_loc(session, method='nearest')]
      whatever_this_mask_is = participant_df.index[nearest_something]
      participant_tau = pd.DataFrame(participant_df['META_TEMPORAL_SUVR'][whatever_this_mask_is])
```

Reply:
I cannot get this to work, I think because I've taken the baseline_df out of the loop, which is the df containing the column session. This is what I have so far. Do you have any ideas? I have tried changing session to baseline_df.session...

```tau_dfs_list = []

for tau_id, participant_df in tau_df.groupby(level='RID'):
    if tau_id in tau_df.index.intersection(baseline_df.index):
        participant_df.set_index('EXAMDATE', inplace=True)
        nearest_something = [participant_df.index.get_loc(session, method='nearest')]```

Reply:
I was a bit confused with this comment as this is referencing L93, and not so relevant to L98.
It's not easy to help you refactor your code without running it.
There's a lot of action chained in one line, changing the property of the objects, so it's very difficult to imagine what the outputs are like. 
If you want interactive code quality feedback, check out sourcery and install it as a vscode plug in. 

This is a sketch. You still need to work out what actually the code should be.

```
grouped_tau_df = tau_df.groupby(level='RID')  # I have no idea what the output should look like, and why it can be iterate through a for loop and unpacked
common_ids = baseline_df.index.intersection(grouped_tau_df.index)

for id in common_ids:
    # use id to look up group in grouped_tau_df
    # use id to look up session in baseline_df
    # Followed by the original code under the if state check. 
```

Reply:
Thank you so much!! The ```nearest_something``` section was referencing Line 98, sorry for the confusion. I have figured it out with your suggestions.